### PR TITLE
[MIG][12.0] mass_mailing_newsletter_welcome_mail

### PR DESCRIPTION
**Pull request by ernestotejeda**
Cc @Tecnativa TT19367

### CODE REVIEW

File: mass_mailing_newsletter_welcome_mail/controllers/main.py [hunk 1]
```diff
@@ -16,16 +16,19 @@ def subscribe(self, list_id, email, **post):
         if not template:
             return result
         # Welcome new subscribers
-        contact = request.env["mail.mass_mailing.contact"].sudo().search([
-            ('list_ids', 'in', list_.ids),
-            ('email', '=', email),
-            ("opt_out", "=", False),
+        contact_obj = request.env["mail.mass_mailing.contact"].with_context(
+            default_list_ids=[list_id])
```
Comment (pedrobaeza) [inline mass_mailing_newsletter_welcome_mail/controllers/main.py:20]: Why this change?

File: mass_mailing_newsletter_welcome_mail/controllers/main.py [hunk 2]
```diff
@@ -16,16 +16,19 @@ def subscribe(self, list_id, email, **post):
...
+            ('is_blacklisted', '=', False),
         ], limit=1)
-        template.with_context(list_name=list_.name).send_mail(
-            contact.id,
-            # Must send now to use context
-            force_send=True,
-            # If we cannot notify, the visitor shouldn't be bothered
-            raise_exception=False,
-        )
+        if contact:
```
Comment (pedrobaeza) [inline mass_mailing_newsletter_welcome_mail/controllers/main.py]: Do this instead for avoiding too much diff:
```python
if not contact:
    return
```

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

**Review by Tardo — APPROVED** [review]
:+1: LGTM

File: mass_mailing_newsletter_welcome_mail/controllers/main.py (hunk 1 above)
Comment (pedrobaeza) [reply mass_mailing_newsletter_welcome_mail/controllers/main.py:20]: Can you please answer about this?

Comment (ernestotejeda) [reply mass_mailing_newsletter_welcome_mail/controllers/main.py:20]: Now in v12 the contacts are not opted out by themselves, but they are opted out or not from a particular list. One way to search contacts that are opted out from a particular list is that, by searching by the opt_oup field of mass.mailing.contact (which is now a calculated field) and passing in the context the list id.
check this
https://github.com/odoo/odoo/blob/fe35870fa0024a1e8adebb386f343f25fdea2c7b/addons/mass_mailing/models/mass_mailing.py#L250
On the other hand, I assumed that so in v11 we didn't want to send mail to contacts that had been opted out from the list, now in v12 that blacklisting exists, we don't whant either to send mail to those who are blacklisted

**Comment by ernestotejeda** [comment]
Changes done

File: mass_mailing_newsletter_welcome_mail/controllers/main.py (hunk 1 above)
Comment (pedrobaeza) [reply mass_mailing_newsletter_welcome_mail/controllers/main.py:20]: Well, if they are subscribing just now, how they are going to be opted-out?

If you keep it (and for the rest of cases), please always explain this not easy to deduce things as a comment in the code.

Comment (ernestotejeda) [reply mass_mailing_newsletter_welcome_mail/controllers/main.py:20]: Maybe in Odoo 11.0 this is not necessary but in v12 yes because the re-subscription of a contact doesn't work well. I made a issue to Odoo for this: https://github.com/odoo/odoo/issues/39604

Comment (pedrobaeza) [reply mass_mailing_newsletter_welcome_mail/controllers/main.py:20]: OK, then please add that as comment:
```python
# Needed for being re-subscribed while odoo/odoo#39604 is not fixed
```

**Review by pedrobaeza — APPROVED** [review]
